app/data_providers/dispatch.py

```diff
@@ -29,22 +30,60 @@
 from app.rankers.algorithms import rank_by_preferred_topics, spread_topics
 
 
-# todo: add thompson sampling
+def _empty_on_error(func):
+    """ Fallback to empty recommendations if Qdrant error occurred """
+    @functools.wraps(func)
+    async def wrapper(*args, **kwargs):
+        try:
+            return await func(*args, **kwargs)
+        except QdrantError:
+            return []
+    return wrapper
+
+
 class Item2ItemDispatch:
-    def __init__(self,
-                 item_recommender: Item2ItemRecommender):
+    def __init__(self, item_recommender: Item2ItemRecommender):
         self.item_recommender = item_recommender
 
-    async def related(self, resolved_id: int, count: int) -> List[CorpusRecommendationModel]:
-        recs = await self.item_recommender.related(resolved_id, count)
-        return [CorpusRecommendationModel(corpus_item=r) for r in recs]
-
+    async def after_save(self, resolved_id: int, count: int) -> List[CorpusRecommendationModel]:
+        try:
+            recs = await self.item_recommender.related(resolved_id, count)
+        except Item2ItemError:
+            # do not fallback for "Similar stores" after saving
+            recs = []
+        return self._sample(recs, count)
+
+    @_empty_on_error
+    async def after_article(self, resolved_id: int, count: int) -> List[CorpusRecommendationModel]:
+        try:
+            recs = await self.item_recommender.related(resolved_id, count)
+        except Item2ItemError:
+            # fallback to frequently saved for "You Might Also Like"
+            recs = await self.item_recommender.frequently_saved(count=100)
```

Comment:
I think it's important to make clear in this function's naming that it only returns curated items, for example:
```suggestion
            recs = await self.item_recommender.curated_frequently_saved(count=100)
```

Reply:
All recommendations provided by this class are currently curated. We are just not allowed to show non-curated stuff. Also, there is `is_syndicated` flag that can make this both curated and syndicated. So I would keep as is.